datums_warehouse/broker/source.py

```diff
@@ -1,22 +1,126 @@
+import gzip
+import json
+import logging
+import time
+from pathlib import Path
+
 import requests
+from more_itertools import first
 
 from datums_warehouse.broker.adapters import KrakenAdapter
-from datums_warehouse.broker.datums import CsvDatums
-from datums_warehouse.broker.validation import validate
+from datums_warehouse.broker.datums import CsvDatums, floor_to_interval
+from datums_warehouse.broker.validation import validate, DataError
+
+LEDGER_FREQUENCY = 6
+logger = logging.getLogger(__name__)
 
 
 def to_nano_sec(t):
-    return t * 1e9
+    return int(t * 1e9)
 
 
-class KrakenSource:
+class KrakenServerTime:
+    _TIME_URL = "https://api.kraken.com/0/public/Time"
+
+    def now(self):
+        res = requests.get(self._TIME_URL).json()
+        return res['result']['unixtime']
+
+
+class KrakenTrades:
     _TRADE_URL = "https://api.kraken.com/0/public/Trades"
 
-    def __init__(self, pair, interval):
+    def __init__(self, cache_dir, pair, max_results):
         self._pair = pair
+        self._cache = Path(cache_dir) / self._pair / "kraken_cache.gz"
+        self._max_results = max_results
+
+    def get(self, since, until):
+        res = self._get_cached(since, until)
+        len_results = get_len(res)
+        while get_last(res) < until and len_results < self._max_results:
+            time.sleep(LEDGER_FREQUENCY)
+            next_trades = self._query_remote_trades(get_last(res) or since)
+            self._update_cache(next_trades)
+            res = combine_trades(res, next_trades)
+            len_results = get_len(res)
+            logger.info(f" <<< received total: {len_results}")
```

Comment:
The way this is done is a bit convoluted. Better would be to query the cash if it is up to date - update it if not and then use the cache for the query.